lib/validators/specValidator.ts

```diff
@@ -219,9 +220,9 @@ export class SpecValidator<T extends CommonValidationResult> {
     const err: CommonError = {
       code: code.name,
       id: code.id,
-      message: message,
-      innerErrors: innerErrors ? innerErrors : undefined
+      message: message
     }
+    setMutableProperty(err, "innerErrors", innerErrors ? innerErrors : undefined)
```

Comment:
having assignment in a separate package is not ok imo! why did i have to go through 5 github pages to see what this does?! please just do assignment normally or copy + paste the code or use a legit 3rd party library like immutablejs

Reply:
I would like to use good typescript utility libraries. Sadly, there are not many of them. Most of the nice libraries, such as `lazy.js` are `javascript` libraries with sometimes broken or incorrect type definitions.

Reply:
I think it would be better to improve the type definitions of libraries that are being maintained than to write our own

Reply:
@billytrend  The challenges are: 
- someone will also need to maintain this type definition because libraries are changed 
- it's really hard to understand from documentation what is the real TypeScript signature of the functions. TypeScript is limited to describe types, especially conditional types. Some libraries have functions with context dependent parameters. For example, if the first parameter is A, then the type of the second parameter should be B.

I'm ok with this if someone will do it :-) I've tried, thanks.

Reply:
@billytrend here's an example. Function `reduce` which is implemented in `lazy.js` and `lodash` in some extend. The problem with this functions, that it must have to TypeScript signatues:

```ts
export const reduce = <T, A>(
    input: Iterable<T>|undefined,
    func: (a: A, b: T, i: number) => A,
    init: A,
): A
```

and

```ts
export const reduce = <T>(
    input: Iterable<T>|undefined,
    func: (a: T, b: T, i: number) => T,
): T|undefined
```

It's possible to describe two signatures for the same function in TypeScript but you are losing all compile-time type checks in this case. So in `ts-common/iterable` I have two functions `fold` and `reduce` with different signatures https://github.com/ts-common/iterator/blob/master/index.ts#L133

Reply:
I don't really see how you lose type safety here but even if this is the case i think the type-safety++ is not worth the readability-- and maintainability--

Reply:
@billytrend I would like to have all, readability, maintainability, consistency, and especially type safety :-) https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Functions.md#overloads